### ISSUE TITLE
incompatible_disable_genrule_cc_toolchain_dependency: remove genrule's implicit dependency on the cc_toolchain

### ISSUE THREAD

**Issue by katre** [bazelbuild/bazel#6867]
Flag: --incompatible_disable_genrule_cc_toolchain_dependency
Available since: 0.22
Will be flipped in: 0.25

Currently, every genrule target has an implicit dependency on the cc_toolchain, caused by the ":cc_toolchain" late-bound attribute. With the recent migration to using the "toolchains" attribute to setting this dependency, the only remaining use of ":cc_toolchains" is to provide the CC_FLAGS Make variable.

In order to remove this dependency, a new rule will be added, `cc_flags_supplier`, that exists only to make the CC_FLAGS Make variable available to other targets.

Once users have migrated to using the new rule to provide CC_FLAGS, the old behavior will be removed. Also, CcToolchainProvider will stop providing the CC_FLAGS Make variable.

Migration notes:
To migrate, add a new toolchains dependency for any `genrule` target that accesses CC_FLAGS.

Example old behavior:
```
genrule(
  name = "display_flags",
  out = ["cc_flags.log"],
  cmd = "echo $(CC_FLAGS) > $@",
  toolchains = ["@bazel_tools//tools/cpp:current_cc_toolchain"],
)
```

New behavior:
```
cc_flags_supplier(name = "cc_flags")
genrule(
  name = "display_flags",
  out = ["cc_flags.log"],
  cmd = "echo $(CC_FLAGS) > $@",
  toolchains = [":cc_flags", "@bazel_tools//tools/cpp:current_cc_toolchain"],
)
```

**Comment by hlopko** [comment]
@aehlig this issue didn't have correct labels, I just added them, FYI.
@katre can you make sure this flag is mentioned in the release notes? Thanks!

**Comment by hlopko** [comment]
I also edited the issue description, the flag is not present in 0.21, only in 0.22. So the flag will be flipped in 0.23.

**Comment by katre** [comment]
Thanks, I'll take care of the release notes.

**Comment by katre** [comment]
Added the link to the release notes for 0.22, let me know if more detail is needed.

**Comment by philwo** [comment]
This flag was not flipped in time for the Bazel 0.23.0 release and will thus be postponed to Bazel 0.24.0.

**Comment by scottmin0r** [comment]
This flag doesn't appear in the output of `bazel help build` - should it?

Output of `bazel help build | grep 'incompatible_disable_'`:
``` 
  --[no]incompatible_disable_expand_if_all_available_in_flag_set (a boolean; default: "false")
  --[no]incompatible_disable_legacy_crosstool_fields (a boolean; default: "false")
  --[no]incompatible_disable_runtimes_filegroups (a boolean; default: "false")
  --[no]incompatible_disable_deprecated_attr_params (a boolean; default: "false")
  --[no]incompatible_disable_objc_provider_resources (a boolean; default: "false")
```

Output of `bazel help build | grep 'incompatible_disable_genrule'`:

```
<empty>
```

Output of `bazel version`:

```
Build label: 0.22.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Mon Jan 28 12:58:08 2019 (1548680288)
Build timestamp: 1548680288
Build timestamp as int: 1548680288
```

**Comment by katre** [comment]
I can change it to be documented, but not sure that can be added to the 0.23 release. I'm not really sure it'd needed: it's a migration flag to help test the requirement to use `@bazel_tools//tools/cpp:cc_flags` instead of automatically getting `$(CC_FLAGS)`.

**Comment by dslomov** [comment]
Why did you remove migration-0.23 and migration-0.22 labels?

**Comment by katre** [comment]
Because I misunderstood the directions. Trying to fix issues now.

**Comment by dslomov** [comment]
Thanks!

**Comment by meteorcloudy** [comment]
TensorFlow still needs migration for this flag. The reason is one of its dependency doesn't work anymore.
```
$ bazel build --nobuild --config=opt tensorflow/tools/pip_package:build_pip_package --incompatible_disable_legacy_cc_provider --incompatible_disable_genrule_cc_toolchain_dependency
INFO: Build option --incompatible_disable_genrule_cc_toolchain_dependency has changed, discarding analysis cache.
ERROR: /usr/local/google/home/pcloudy/.cache/bazel/_bazel_pcloudy/d2f15366e3b15100d3593f2d595721a3/external/com_github_googlecloudplatform_google_cloud_cpp/google/cloud/BUILD:31:11: in cmd attribute of genrule rule @com_github_googlecloudplatform_google_cloud_cpp//google/cloud:generate_build_info: $(CC) not defined
ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted: Analysis of target '@com_github_googlecloudplatform_google_cloud_cpp//google/cloud:generate_build_info' failed; build aborted
```

To reproduce
```
git clone https://github.com/googleapis/google-cloud-cpp.git
cd google-cloud-cpp/
bazel build //google/cloud:generate_build_info  --incompatible_disable_genrule_cc_toolchain_dependency
```

**Comment by katre** [comment]
Yes, I am working to upgrade google-cloud-cpp (internal change not exported yet). Once that is done and they release I will update TF.

**Comment by katre** [comment]
Unfortunately I think this means I can't flip this flag before 0.25 is cut, unless the process goes surprisingly fast in the next day or so.

**Comment by meteorcloudy** [comment]
Thanks for doing the fix!

**Comment by hlopko** [comment]
I wouldn't dismiss the option of disabling TF on downstream for a few days and flip this flag tomorrow. We know they will be migrated, we have the fix, so there is no risk of unforseen bugs.

I'd take this option :) And @meteorcloudy said he'd be fine with it too.

**Comment by katre** [comment]
Re-opening until the flag is verified and the code is removed.

**Comment by katre** [comment]
Re-closing because I still don't understand the incompatible flag workflow.
